peps/pep-0747.rst

```diff
@@ -0,0 +1,903 @@
+.. _isa: https://github.com/xdslproject/xdsl/blob/ac12c9ab0d64618475efb98d1d197bdd79f593c3/xdsl/utils/hints.py#L23
+
+-  Converters:
+
+   -  If a value is assignable to (or coercible to) a type expression
+      then returns the value narrowed to (or coerced to) that type
+      expression. Otherwise raises an exception.
+   -  Pattern 1:
+      ``def convert[T](value: object, typx: TypeExpr[T]) -> T``
+
+      -  Examples: cattrs.BaseConverter.\ `structure`_, trycast.\ `checkcast`_,
+         typedload.\ `load`_
+
+   -  Pattern 2:
+
+      ::
+
+        class Converter[T]:
+            def __init__(self, typx: TypeExpr[T]) -> None: ...
+            def convert(self, value: object) -> T: ...
+
+      -  Examples: pydantic.\ `TypeAdapter(T).validate_python`_,
+         mashumaro.\ `JSONDecoder(T).decode`_
+
+.. _structure: https://github.com/python-attrs/cattrs/blob/5f5c11627a7f67a23d6212bc7df9f96243c62dc5/src/cattrs/converters.py#L332-L334
+.. _checkcast: https://github.com/davidfstr/trycast#checkcast-api
+.. _load: https://ltworf.github.io/typedload/
+.. _TypeAdapter(T).validate_python: https://stackoverflow.com/a/61021183/604063
+.. _JSONDecoder(T).decode: https://github.com/Fatal1ty/mashumaro?tab=readme-ov-file#usage-example
+
+-  Typed field definitions:
+
+   -  Pattern:
+
+      ::
+
+        class Field[T]:
+            value_type: TypeExpr[T]
+
+   -  Examples: attrs.\ `make_class`_, dataclasses.\ `make_dataclass`_, `openapify`_
+
+.. _make_class: https://www.attrs.org/en/stable/api.html#attrs.make_class
+.. _make_dataclass: https://github.com/python/typeshed/issues/11653
+.. _openapify: https://github.com/Fatal1ty/openapify/blob/c8d968c7c9c8fd7d4888bd2ddbe18ffd1469f3ca/openapify/core/models.py#L16
+
+The survey also identified some introspection functions that take
+annotation expressions as input using plain ``object``\ s which would
+*not* gain functionality by marking those inputs as ``TypeExpr[]``:
+
+-  General introspection operations:
+
+   -  Pattern: ``def get_annotation_info(maybe_annx: object) -> object``
+   -  Examples: typing.{`get_origin`_, `get_args`_},
+      `typing_inspect`_.{is_*_type, get_origin, get_parameters}
+
+.. _get_origin: https://docs.python.org/3/library/typing.html#typing.get_origin
+.. _get_args: https://docs.python.org/3/library/typing.html#typing.get_args
+.. _typing_inspect: https://github.com/ilevkivskyi/typing_inspect?tab=readme-ov-file#readme
+
+
+Specification
+=============
+
+A ``TypeExpr`` represents a `type expression`_ such as ``Optional[str]``,
+``Union[int, str]``, or ``MyTypedDict``. A ``TypeExpr`` type is written as
+``TypeExpr[T]`` where ``T`` is a type or a type variable. It can also be
+written without brackets as just ``TypeExpr``, in which case a type
+checker should apply its usual type inference mechanisms to determine
+the type of its argument, possibly ``Any``.
+
+Type expressions include more values than `annotation expressions`_ but
```

Comment:
For the benefit of readers of the PEP, it may be worth recapitulating what a "type expression" is. I think that may fit well in a "Rationale" section, which is about justifying why we're using this precise approach.